### PR TITLE
DOCSP-25900: recursive type support

### PR DESCRIPTION
**Pull request by rustagir**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-java/blob/master/REVIEWING.md)

JIRA - https://jira.mongodb.org/browse/DOCSP-25900
Staging - https://docs-mongodbcom-staging.corp.mongodb.com/java/docsworker-xlarge/DOCSP-25900-recursive-type-support/fundamentals/data-formats/document-data-format-record/#operations-with-recursive-types

## Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?

### CODE REVIEW

File: source/fundamentals/data-formats/document-data-format-record.txt [hunk 1]
```diff
@@ -191,3 +190,38 @@ as shown in the following code:
 
       NetworkDeviceRecord[deviceId=fedc..., name=Enterprise Wi-fi, deviceType=router]
 
+Operations with Recursive Types
+-------------------------------
+
+The driver natively supports encoding and decoding cyclical
+references without causing runtime recursion. The following code
+provides an example of a recursive class design:
```
Comment (mongoKart) [inline source/fundamentals/data-formats/document-data-format-record.txt:198]: Issue: My Java's a little rusty, but isn't this a method named RecordTree that returns a record? If not, please ignore :)

Suggestion: Change code sample to a recursive class.

**Review by mongoKart — CHANGES_REQUESTED** [review]
One question. Also double-checking you're sending to technical review at some point so I don't have to scrutinize code samples

File: source/fundamentals/data-formats/document-data-format-record.txt [hunk 2]
```diff
@@ -191,3 +190,38 @@ as shown in the following code:
 
       NetworkDeviceRecord[deviceId=fedc..., name=Enterprise Wi-fi, deviceType=router]
 
+Operations with Recursive Types
+-------------------------------
+
+The driver natively supports encoding and decoding cyclical
+references without causing runtime recursion. The following code
```
Comment (jyemin) [inline source/fundamentals/data-formats/document-data-format-record.txt:197]: It's not the references that are cyclical.  The driver doesn't actually handle that.  It's when there are cycles in the type definitions.  In this case, RecordTree has two components of type RecordTree.  So there is a (short) cycle in the type definitions.  But the support generalizes to any sort of cycle in the type system, e.g.  RecordOne with a component of type RecordTwo with a component of type RecordThree with a component of type RecordOne.

**Review by jyemin — CHANGES_REQUESTED** [review]

File: source/fundamentals/data-formats/document-data-format-record.txt (hunk 2 above)
Comment (rustagir) [reply source/fundamentals/data-formats/document-data-format-record.txt:197]: do you think that I should expand the example to add a second (or third) record definition to fully illustrate support for cycles? I will also reword.

Comment (jyemin) [reply source/fundamentals/data-formats/document-data-format-record.txt:197]: Probably not necessary, I was just trying to explain it to you.  But I don't think we should use the phrase "cyclical references" as that will most likely be interpreted as references of instances, not types.

File: source/fundamentals/data-formats/document-data-format-record.txt (hunk 1 above)
Comment (jyemin) [reply source/fundamentals/data-formats/document-data-format-record.txt:198]: This feature is about records, not classes, so I think it should stay as is.

**Review by jyemin — APPROVED** [review]
LGTM

**Review by mongoKart — APPROVED** [review]
LGTM!

**Pull request merged**